k2/python/csrc/torch/ragged.cu

```diff
@@ -0,0 +1,95 @@
+/**
+ * @brief python wrappers for Ragged<T>.
+ *
+ * @copyright
+ * Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <vector>
+
+#include "k2/csrc/ragged.h"
+#include "k2/python/csrc/torch/ragged.h"
+#include "k2/python/csrc/torch/torch_util.h"
+#include "torch/extension.h"
+
+namespace k2 {
+
+static void PybindRaggedShape(py::module &m) {
+  using PyClass = RaggedShape;
+  py::class_<PyClass> pyclass(m, "RaggedShape");
+
+  // NOTE: we do not provide constructors of RaggedShape
+  // in Python, so the only way to get an instance of
+  // RaggedShape is from Ragged<T>.
+
+  pyclass.def("dim0", &PyClass::Dim0);
+  pyclass.def("total_size", &PyClass::TotSize, py::arg("axis"));
+
+  // TODO(fangjun): Append has not been implemented yet
+  // pyclass.def("append", &PyClass::Append, py::arg("other"));
+
+  pyclass.def("numel", &PyClass::NumElements);
+
+  pyclass.def(
+      "row_splits",
+      [](PyClass &self, int32_t axis) -> torch::Tensor {
+        Array1<int32_t> &row_splits = self.RowSplits(axis);
+        return ToTensor(row_splits);
+      },
+      py::keep_alive<0, 1>());
+
+  pyclass.def(
+      "row_ids",
+      [](PyClass &self, int32_t axis) -> torch::Tensor {
+        Array1<int32_t> &row_ids = self.RowIds(axis);
+        return ToTensor(row_ids);
+      },
+      py::keep_alive<0, 1>());
+
+  pyclass.def("num_axes", &PyClass::NumAxes);
+  pyclass.def("max_size", &PyClass::MaxSize, py::arg("axis"));
+  pyclass.def("index", &PyClass::Index, py::arg("axis"), py::arg("i"));
+  pyclass.def(
+      "get_offset",
+      [](PyClass &self, const std::vector<int32_t> &indexes) -> int32_t {
+        return self[indexes];
+      });
+}
+
+template <typename T>
+static void PybindRaggedTpl(py::module &m, const char *name) {
+  using PyClass = Ragged<T>;
+  py::class_<PyClass> pyclass(m, name);
+
+  pyclass.def_readwrite("shape", &PyClass::shape);
+  pyclass.def_readwrite("values", &PyClass::values);
+
+  pyclass.def("num_axes", &PyClass::NumAxes);
+  pyclass.def("index", &PyClass::Index, py::arg("axis"), py::arg("i"));
+  pyclass.def("remove_axis", &PyClass::RemoveAxis, py::arg("axis"));
+
+  pyclass.def(
+      "cuda",
```

Comment:
OK..  this will probably work given the current compilation setup, where we use only the PyTorch context if we're compiled with pytorch.  When I added the member-function GetCpuContext() to the Context class I assumed we would be generating these contexts from existing contexts, mostly.  I guess this is OK, but I just want to remind people that my original idea was that we'd mostly generate contexts from other contexts.

Also I don't know which GPU this would give us in cases where there were multiple GPUs ?

Reply:
> Also I don't know which GPU this would give us in cases where there were multiple GPUs ?

This would be the GPU specified by the `gpu_id`. If no `gpu_id` is specified.
PyTorch will use the default GPU, like the following code:

```python
a = torch.arange(10).cuda()  // PyTorch select the `default gpu`.
```

> When I added the member-function GetCpuContext() to the Context class I assumed we would be generating these contexts from existing contexts

How do we get the first context?  I think there are only allocators associated with
`torch::Tensor`, but no `Context`.

Reply:
Mm, I hadn't mapped out the details, e.g. I don't know whether you can get this from the C++ or have to do it from Python.  Anyway the way it is now seems fine.  We could perhaps just remove that function GetCpuContext() from the Context interface.